packages/salesforcedx-vscode-core/src/commands/functions/forceFunctionStart/ForceFunctionContainerStartExecutor.ts

```diff
@@ -0,0 +1,164 @@
+
+    const handleError = (error: any) => {
+      functionDisposable.dispose();
+      let unexpectedError = true;
+      (Object.keys(
+        forceFunctionStartErrorInfo
+      ) as ForceFunctionStartErrorType[]).forEach(errorType => {
+        const {
+          cliMessage,
+          errorNotificationMessage
+        } = forceFunctionStartErrorInfo[errorType];
+        if (error.text?.includes(cliMessage)) {
+          unexpectedError = false;
+          telemetryService.sendException(errorType, errorNotificationMessage);
+          notificationService.showErrorMessage(errorNotificationMessage);
+          channelService.appendLine(errorNotificationMessage);
+        }
+      });
+
+      if (unexpectedError) {
+        const errorNotificationMessage = nls.localize(
+          this.UNEXPECTED_ERROR_KEY
+        );
+        telemetryService.sendException(
+          this.UNEXPECTED_ERROR_KEY,
+          errorNotificationMessage
+        );
+        notificationService.showErrorMessage(errorNotificationMessage);
+        channelService.appendLine(errorNotificationMessage);
+      }
+      channelService.showChannelOutput();
+    };
+
+    this.functionsBinary.on(BINARY_EVENT_ENUM.PACK, writeMsg);
+    this.functionsBinary.on(BINARY_EVENT_ENUM.CONTAINER, writeMsg);
+
+    this.functionsBinary.on(BINARY_EVENT_ENUM.LOG, (msg: any) => {
+      if (msg.level === 'debug') return;
+      if (msg.level === 'error') {
+        handleError(msg);
+      }
+
+      if (msg.text) {
+        writeMsg(msg);
+      }
+
+      // evergreen:benny:message {"type":"log","timestamp":"2021-05-10T10:00:27.953248-05:00","level":"info","fields":{"debugPort":"9229","localImageName":"jvm-fn-init","network":"","port":"8080"}} +21ms
+      if (msg.fields && msg.fields.localImageName) {
+        channelService.appendLine(`'Running on port' :${msg.fields.port}`);
+        channelService.appendLine(
+          `'Debugger running on port' :${msg.fields.debugPort}`
+        );
+      }
+    });
+    // Allows for showing custom notifications
+    // and sending custom telemtry data for predefined errors
+    this.functionsBinary.on(BINARY_EVENT_ENUM.ERROR, handleError);
+  }
+  public async cancelFunction(
+    registeredStartedFunctionDisposable: vscode.Disposable
+  ): Promise<void> {
+    if (this.functionsBinary !== undefined) {
+      this.functionsBinary.cancel();
+    }
+    this.functionsBinary = undefined;
+    registeredStartedFunctionDisposable.dispose();
+  }
+
+  public async buildFunction(
+    functionName: string,
+    functionDirPath: string
+  ): Promise<void> {
+    channelService.appendLine(`Building ${functionName}`);
+
+    if (this.functionsBinary === undefined) {
+      throw new Error('Unable to find binary for building function.');
+    }
+
+    await this.functionsBinary.build(functionName, {
+      verbose: true,
+      path: functionDirPath
+    });
+  }
+
+  public async startFunction(
+    functionName: string,
+    functionDirPath: string
```

Comment:
`functionDirPath` isn't used.  Can it be removed?

Reply:
good call. in the abstract method def the param is optional I can just exclude it (needed for containerless implementation)